[Curtis Chang Kai Xian] iP

curtischang2510 opened the pull request:
## Blitz 
> "Your mind is for having ideas, not holding them" - David Allen [(source)](https://www.google.com.sg/)

Blitz frees your mind of having to remember things you need to do. It's

- Test-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use

All you need to do is, 
1. Download it from [here](https://www.google.com.sg/)
2. double clicl it
3. add your tasts 
4. let it manage your tasks for you

And it is **FREE!**

Features: 
- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)


If you are a Java programmer, you can use it to practice Java too. Here's the `main` method: 

``` Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

curtischang2510 committed:
Add: Blitz.java

Renamed Chatbot from Duke to Blitz.

curtischang2510 committed:
Add user input functionality in Blitz.java

Accepts user input and echos input. Terminates when user types "bye" (case insensitive)

curtischang2510 committed:
Revert "Add user input functionality in Blitz.java"

This reverts commit 78f52167dd513642d0fcc98d5035c4982e69ee28.

curtischang2510 committed:
bug fix: code used to still echo "bye" before printing "till we meet again..."

Changed the while loop condition. Trim trailing white space before checking

curtischang2510 committed:
Add userInput recording

Tracks user input history and prints it in list format when user inputs "list" (case insensitive)

curtischang2510 committed:
Add Task class

Strings stored in an array list

By encapsulating information in Task object we do not have to create multiple array lists to store information for same Task instance

curtischang2510 committed:
Add functions to abstract logic within while loop

While loop used to handle all the logic checking

Flow is less clear when the logic is being handled within main function.  After the change main is more readable

curtischang2510 committed:
Add mark and unmark feature for tasks

Tasks can now be marked as completed or uncompleted. Error checking present for when an non-integer input is given

curtischang2510 committed:
Bug fix: ignore when user inputs an empty line

If user types only spaces or returns immediately, that is taken as an entry

Users can accidentally press enter, we do not want to record empty lines as a valid task

curtischang2510 committed:
Add 3 Task subclasses

All tasks were considered to be the same type

User can input different kinds of tasks with different information

curtischang2510 committed:
Add factory method

Creation of new Task() handled in main

Makes code main() more tidy. Factory method can determine the correct
task subclass based on the entire description

curtischang2510 committed:
Add check for invalid task type

With task subclasses users can potentially key in invalid task types.
Factory method returns null when invalid type is given.

curtischang2510 committed:
Task toString(): remove the task type from toString

Printed the task type at the start before description

Task type is now represented by the letter on the extreme left

curtischang2510 committed:
Add test

test output when 1 task of each type is added

curtischang2510 committed:
Add InvalidTaskNameException & NoTaskDescriptionException

Increases readability of exception handling

curtischang2510 committed:
Add exception handling feature

Invalid task exception was caught when factory method returns null

Factory method explicitly throws an InvalidTaskException.
Factory method throws exception when task type is not followed by
description. Improves readability of code and exception handle is more
appropriate than if-else blocks in this scenario

curtischang2510 committed:
Add delete feature

User can delete a task, regardless of whether it is completed

curtischang2510 committed:
Add enum TaskType

Clearer management for different types of tasks. Improves readability

curtischang2510 committed:
style: Add whitespace

curtischang2510 committed:
Update methods to be non-static

All the methods were static, only accommodating for 1 instance of Blitz

Changed implementation to allow multiple instances of chatbot by making
it non-static. Removed logic from main, moved to start()

curtischang2510 committed:
Remove inputHistory argument

inputHistory passed as arguments into multiple functions

Every instance has its own history, methods reference that directly

curtischang2510 committed:
Remove inputHistory argument

inputHistory passed as arguments into multiple functions

Every instance has its own history, methods reference that directly

curtischang2510 committed:
Add save and load feature for task persistence

- Program data was lost upon termination.
- No functionality existed to save tasks to disk or load them on startup.

Lets,
- Implement reading from blitz.txt on program startup to load tasks.
- Implement writing to blitz.txt when the program terminates to save
tasks.
- Ensure that program data persists after exiting, allowing tasks to be
retained across sessions.

curtischang2510 committed:
Add support for LocalDate & LocalTime

Dates are taken as strings

Lets,
- Specify a date format for clients to follow and store them as LocalDate
and LocalTime instances
- Display date and time in different formats
- Allow for easy editing of task time and date (feature not implemented
yet)

curtischang2510 committed:
Add Storage, TaskList, Parser, Ui classes

Logic mostly handled by Task and Blitz classes

Lets,
- Implement Storage class to handle data persistence (read/write to
blitz.txt)
- Add Parser class to process and understand user commands
- Creat TaskList class to manage tasks (add, delete, mark/unmark tasks)
- Introduce UI class to handle user interaction and console output
- Improve overall design by separating concerns and modularizing code
- Make the code easier to scale to different task types

curtischang2510 committed:
Remove unused imports

curtischang2510 committed:
Merge branch 'add-gradle-support'

curtischang2510 committed:
Add Gradle for automating build tasks

- Integrated Gradle into the project to streamline build processes.
- Set up basic Gradle tasks for checkstyle, testing, and running the
application.
- Added dependencies and configurations to automate project workflows.

curtischang2510 committed:
Add JUnit test cases for Blitz and Task classes

- Added tests for `actionBasedOnInput` method in `Blitz` class to handle
various input scenarios including empty input, invalid indexes, and
non-numeric values.

- Added tests for `createTask` method in `Task` class to validate
exception handling for empty descriptions

curtischang2510 committed:
Merge branch 'branch-level-8'

* branch-level-8:
  Add support for LocalDate & LocalTime

# Conflicts:
#	src/main/java/Task.java

curtischang2510 committed:
Add Task.Task import statements

curtischang2510 committed:
Merge branch 'add-gradle-support' into branch-A-MoreOOP

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

curtischang2510 committed:
Update build.grade set mainClass as Blitz.java

curtischang2510 committed:
Merge tag 'A-JUnit'

* tag 'A-JUnit':
  Add JUnit test cases for Blitz and Task classes
  Add Gradle for automating build tasks

# Conflicts:
#	build.gradle
#	src/main/java/Blitz.java
#	src/main/java/Task/Task.java

curtischang2510 committed:
Add unit test for command handling

Unit test based on old commit (before MoreOOP)

- Update unit test due to change in implementation

curtischang2510 committed:
Add Find feature

User could not filter tasks by description

Lets,
- allow users to find tasks by searching for keywords in description

curtischang2510 committed:
Add time format flexibility

User had to input time with a colon

Lets,
- Allow user to just type in 24 hour format (e.g. 1800)

curtischang2510 committed:
Add JavaDocs

- Add docs to AddTaskCommand, DeleteCommand, ExitCommand, ListCommand,
MarkCommand, UnmarkCommand, Parser, Task, TaskList, Deadline, Event,
Todo

curtischang2510 committed:
Update style

- Format files to adhere to course coding style conventions

curtischang2510 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDocs

# Conflicts:
#	src/main/java/Parser/Parser.java
#	src/main/java/Task/TaskList.java
#	src/main/java/Task/TaskType/Deadline.java

curtischang2510 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Update style

# Conflicts:
#	src/main/java/Blitz.java
#	src/main/java/Commands/DeleteCommand.java
#	src/main/java/Commands/ExitCommand.java
#	src/main/java/commands/AddTaskCommand.java
#	src/main/java/commands/FindCommand.java
#	src/main/java/commands/ListCommand.java
#	src/main/java/commands/MarkCommand.java
#	src/main/java/parser/Parser.java
#	src/main/java/task/Task.java
#	src/main/java/task/TaskList.java
#	src/main/java/task/tasktype/Deadline.java
#	src/main/java/task/tasktype/Event.java
#	src/test/java/CommandTest.java

curtischang2510 committed:
Update style

- Updated style to match course coding conventions

ossaris56 commented on src/main/java/commands/ExitCommand.java:
```diff
@@ -0,0 +1,12 @@
+package commands;
+
+import task.TaskList;
+
```
Missing JavaDoc

ossaris56 commented on src/main/java/commands/ExitCommand.java:
```diff
@@ -0,0 +1,12 @@
+package commands;
+
+import task.TaskList;
+
+public class ExitCommand extends Command {
+
```
Missing JavaDoc

ossaris56 commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,21 @@
+
+import task.Task;
+import task.TaskList;
+
+public class FindCommand extends Command {
+    private final String ITEM_TO_FIND;
+
+    public FindCommand(String item) {
+        this.ITEM_TO_FIND = item;
+    }
+    @Override
+    public void execute(TaskList taskList) {
```
Missing JavaDoc

ossaris56 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,158 @@
+            return "deadline " + description + " /by " + inputDateFormat;
+        }
+    }
+
```
Missing JavaDoc

ossaris56 commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,158 @@
+            throw new InvalidTaskException();
+        }
+    }
+
```
Missing JavaDoc

ossaris56 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.io.File;
+import java.io.IOException;
+
```
Missing JavaDoc

ossaris56 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,37 @@
+    public Storage(File storageFile) {
+        this.storageFile = storageFile;
+    }
+
```
Missing JavaDoc

ossaris56 commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+                + task));
+        System.out.println("---------------\n");
+    }
+
```
Missing JavaDoc

ossaris56 commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,119 @@
+     * @param description A string containing the task description and any associated time details.
+     * @return A new Task object of the appropriate type (Todo, Deadline, or Event).
+     * @throws InvalidTaskException if the task type is invalid.
+     * @throws NoTaskDescriptionException if the task description is missing
```
Missing fullstop

ossaris56 commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * Loads task data from storage and populates the task list.
+     * Done once upon program initialisation
```
Missing fullstop

ossaris56 commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * Writes the current list of tasks to the storage file.
+     * Done after every {@code Command} except {@code ListCommand}
```
Missing fullstop

ossaris56 commented on src/main/java/task/tasktype/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+    protected LocalTime deadlineTime;
+
+    /**
+     * Constructs a new {@code deadline} task with the specified description and deadline
+     *
+     * @param description A description of the deadline task
+     * @param by date that the task needs to be finished by, given in {@code d/M/yyyy} format (e.g. 25/10/2015)
```
Missing fullstops

ossaris56 commented on src/main/java/task/tasktype/Event.java:
```diff
@@ -0,0 +1,59 @@
+    private LocalTime endTime;
+
+    /**
+     * Constructs a new {@code deadline} task with the specified description and deadline
+     *
+     * @param description A description of the deadline task
+     * @param start date that the task needs to be finished by, given in {@code d/M/yyyy} format (e.g. 25/10/2015)
+     * @param end
```
Missing fullstops

ossaris56 reviewed (commented):
LGTM! I think your code is easy to read, just some minor JavaDoc violations.

curtischang2510 committed:
Add GUI

Refactor Blitz from CLI to GUI using JavaFX
- Implement initial GUI layout while retaining CLI functionality
- Add JavaFX components to replicate existing command-line interactions
- Maintain core features and user flow for a smooth transition to GUI
- Update event handling and user input/output processing for GUI-based
interactions

curtischang2510 committed:
Add assert statements to improve code validation

- Inserted assert statements to verify assumptions and ensure proper
functioning of key components.
- These checks will help catch potential issues early during development
and debugging.
- Ensures that input values and internal states are valid before
proceeding with execution.

curtischang2510 committed:
Refactor code to reduce magic strings and improve method readability

- Moved all magic strings to its own class ErrorMessageHandler, improve
maintainability and reduce potential errors
- Refactored long to smaller, more manageable functions
- Removed unused variables and imports

curtischang2510 committed:
Refactor code to use Java Streams

- Improve efficiency be introducing lazy operations

curtischang2510 committed:
Merge pull request #1 from curtischang2510/A-Assertions

Add assert statements to improve code validation

curtischang2510 committed:
Merge branch 'master' into A-CodeQuality

* master:
  Add assert statements to improve code validation

# Conflicts:
#	src/main/java/components/MainWindow.java
#	src/main/java/io/Parser.java
#	src/main/java/task/tasktype/Event.java

curtischang2510 committed:
Merge pull request #2 from curtischang2510/A-CodeQuality

Refactor code to reduce magic strings and improve method readability

curtischang2510 committed:
Merge pull request #3 from curtischang2510/A-Streams

Refactor code to use Java Streams

curtischang2510 committed:
Add tagging feature to tasks

- Implemented tagging functionality for tasks including todos, events,
and deadlines.
- Users can now add tags using the `-t` flag followed by tags.
- Updated task input commands to support optional tags for better
task organization.
- Users can also filter by tags using `find -t (tag names)`
- Example commands for adding tags:
  - `todo math hw -t urgent difficult`
  - `find -t urgent` to filter tasks by tags.

curtischang2510 committed:
Add CI configuration file

- Added .yml file to set up continuous integration

curtischang2510 committed:
Merge branch 'branch-Tagging'

* branch-Tagging:
  Add tagging feature to tasks

curtischang2510 committed:
bug fix: prevent default empty string tag from being added when no tags
are specified

Empty string tag ("") added by default even when no tags were provided.

Lets:
Ensure tags are only added when specified by the user.

curtischang2510 committed:
Add background colours, resizable window, and circular profile pics

- Added background color to the speech bubbles for better visual appeal
- Made the application window resizable to improve user experience
- Updated icons to have a circular shape for a modern look

curtischang2510 committed:
Bug Fix: Correctly identify and retain existing user tags during data
import

Previously, the parser could only identify tags from user input.

Lets,
- Allows tags to be identified and retained when data is imported.

curtischang2510 committed:
docs: Add UI screenshot to documentation

- Included a screenshot of the current UI for reference.

curtischang2510 committed:
docs: Add Javadocs for multiple classes

- Improved clarity by including descriptions for public methods,
constructors, and fields

curtischang2510 committed:
bug fix: Ensure mark/unmark operations persist to storage

mark/unmark operations did not write to storage,
causing changes to task status to be non-persistent.

- Call `writeToStorage()` after every mark/unmark operation to ensure
task status is properly updated and saved.

curtischang2510 committed:
docs: Update README with detailed usage instructions and UI image

- Added detailed instructions for using Blitz, including adding todos,
deadlines, and events.
- Described how to use tags and filter tasks by description or tags.
- Included examples for marking tasks as done/undone and deleting tasks.
- Updated exit instructions and clarified data-saving behavior.
- Added a UI image to enhance user understanding.

curtischang2510 committed:
refactor: Remove unnecessary print statements and add Javadocs

- Removed redundant print statements to clean up the codebase.
- Added Javadocs to improve code documentation and maintainability for
one file.

curtischang2510 committed:
Update README.md

curtischang2510 committed:
Update README.md

curtischang2510 committed:
Update README.md

- set width of image to 300px

curtischang2510 committed:
Update README.md

- set image width to 300px using HTML

curtischang2510 committed:
Refactor: Abstracted switch statement bodies into individual methods

- Moved the logic for handling 'todo', 'event', and 'deadline' from the
switch statement in `dataInputToUserInput` into their own methods
(`todoInputToUserInput`, `eventInputToUserInput`,
`deadlineInputToUserInput`).
- Improved code readability and maintainability by encapsulating each
task type's conversion logic in dedicated methods.
- This change makes it easier to add or modify task type parsing
logic in the future.

curtischang2510 committed:
Chore: Refactor code to adhere to course checkstyle

- Updated code formatting, spacing, and import order to meet the course
checkstyle requirements.
- Ensured compliance with all specified style guidelines to maintain
code quality and uniformity.

curtischang2510 committed:
Merge branch 'master' of github.com:curtischang2510/ip

* 'master' of github.com:curtischang2510/ip:
  Update README.md
  Update README.md
  Update README.md
  Update README.md

# Conflicts:
#	docs/README.md

curtischang2510 committed:
Add and update unit tests for Task class

- Added new unit tests to validate task creation with and without tags.
- Updated existing test cases for creating tasks without descriptions
and tasks with invalid dates or task types.
- Ensured proper coverage for edge cases in task creation scenarios.

curtischang2510 committed:
Fix error: Revert mistakenly pasted TaskTest code into CommandTest class

- Removed TaskTest methods that were incorrectly pasted into the
CommandTest class.
- Restored the original CommandTest methods for proper unit test coverage.

curtischang2510 committed:
Refactor: Updated test structure to adhere to course coding style

- Reorganised import statements to match coding style guidelines
- Improved readability and consistency across test methods

curtischang2510 committed:
Add Javadocs to core classes and methods

Added comprehensive Javadocs to the following files:
- Blitz.java
- Command.java
- Parser.java
- Ui.java
- Tag.java
- Event.java
- TaskType.java

This improves code readability and provides clear documentation for
future developers.

curtischang2510 committed:
refactor: Update application entry point

- Changed the entry point of the application from `Blitz` to `Main`.
This update ensures proper initialization and execution flow through the
new `Main` class.

curtischang2510 committed:
fix: Add launcher class to initialize JavaFX application

- Created a separate launcher class to bypass JavaFX runtime issue.
- Ensures proper initialization of the JavaFX application from the JAR
file.
- Resolves "JavaFX runtime components missing" error.

curtischang2510 committed:
fix: Resolve JavaFX runtime components issue and move data folder to
resources

- Added changes to fix the "JavaFX runtime components missing" error.
- Moved the `data` folder to the `resources` directory for better
project structure.
- Updated file paths to load data correctly from the `resources` folder.

curtischang2510 committed:
Refactor: Removed unused import statements

- Removed `java.io.InputStream` as it was not being utilized in the
codebase. This cleanup improves code readability and adheres to best
practices.

curtischang2510 committed:
docs: Update README in home directory

- Moved and updated the README file to the home directory.

curtischang2510 committed:
Refactor: Empty input and output files for IO redirection tests

- Switch to JUnit for more comprehensive testing

curtischang2510 committed:
Fix: Modify javac command to compile files from subdirectories

- Updated runtest.sh to recursively compile all Java files from
src/main/java and its subdirectories.
- Fixed issue with missing compilation of files in subdirectories
during CI build.

curtischang2510 committed:
Fix: Correct file path typo by adding missing 's' to 'resources'
directory

- Updated file path in the script from 'resource' to 'resources'.
- This resolves issues related to incorrect directory references during
file access.

curtischang2510 committed:
Refactor batch script to recursively compile all Java files

- Updated the Windows batch script to use a `for /r` loop for
recursively compiling all Java files under `src/main/java`.
